packages/react-router/lib/router/router.ts

```diff
@@ -4940,91 +4932,65 @@ async function defaultDataStrategyWithMiddleware(
     return defaultDataStrategy(args);
   }
 
-  // Determine how far down we'll be loading so we only run middleware to that
-  // point.  This prevents us from calling middleware below an action error
-  // boundary below which we don't run loaders
-  let lastIndex = args.matches.length - 1;
-  for (let i = lastIndex; i >= 0; i--) {
-    if (args.matches[i].shouldLoad) {
-      lastIndex = i;
-      break;
-    }
-  }
-
-  let results = await runMiddlewarePipeline(
+  return runMiddlewarePipeline(
     args,
-    lastIndex,
     false,
-    async (keyedResults: Record<string, DataStrategyResult>) => {
-      Object.assign(keyedResults, await defaultDataStrategy(args));
-    },
-    (e, keyedResults) => {
-      // we caught an error running the middleware, copy that overtop any
-      // non-error result for the route
-      Object.assign(keyedResults, {
-        [e.routeId]: { type: "error", result: e.error },
-      });
-    }
-  );
-  return results as Record<string, DataStrategyResult>;
+    () => defaultDataStrategy(args),
+    (e) => ({ [e.routeId]: { type: "error", result: e.error } })
+  ) as Promise<Record<string, DataStrategyResult>>;
 }
 
 type MutableMiddlewareState = {
-  keyedResults: Record<string, DataStrategyResult>;
+  handlerResult: unknown;
   propagateResult: boolean;
 };
 
-export async function runMiddlewarePipeline(
-  {
-    request,
-    params,
-    context,
-    matches,
-  }: (
+export async function runMiddlewarePipeline<T extends boolean>(
+  args: (
     | LoaderFunctionArgs<unstable_RouterContextProvider>
     | ActionFunctionArgs<unstable_RouterContextProvider>
   ) & {
+    // Don't use `DataStrategyFunctionArgs` directly so we can we reduce these
+    // back from `DataStrategyMatch` to regular matches for use in the staticHandler
     matches: AgnosticDataRouteMatch[];
   },
-  lastIndex: number,
-  propagateResult: boolean,
-  handler: (results: Record<string, DataStrategyResult>) => unknown,
-  errorHandler: (
-    error: MiddlewareError,
-    results: Record<string, DataStrategyResult>
-  ) => unknown
+  propagateResult: T,
+  handler: () => T extends true
+    ? MaybePromise<Response>
+    : MaybePromise<Record<string, DataStrategyResult>>,
+  errorHandler: (error: MiddlewareError) => unknown
```

Comment:
I stopped passing a mutable `keyedResults` through to the handler and now you just return one from your handler` directly and we propagate it back out from `runMiddleware` - this should make our user-facing `runMiddleware` function we plan to pass to `dataStrategy` easier